cmd/podman/containers/create.go

```diff
@@ -138,6 +138,10 @@ func create(cmd *cobra.Command, args []string) error {
 		cliVals.InitContainerType = initctr
 	}
 
+	if cmd.Flags().Changed("pod") && strings.HasPrefix(cmd.Flag("pod").Value.String(), "new:") && cmd.Flags().Changed("restart") {
+		return fmt.Errorf("must specify policy for --restart during pod creation; containers in a pod cannot have individual restart policies")
```

Comment:
That is a too big hammer breaking backwards compatibility.  Existing workloads with individual restart policies of containers inside a pod must continue to work.  If we want to restrict it, the check should happen in the backend (probably specgen) to account for REST calls.

I didn't read the full PR yet but at the moment I do not understand the restriction.  The restart policy of a container may be a different one than for the entire pod.

Reply:
Is this still considered breaking backwards compatibility? We agreed on following k8s and in that case the restart policy can only be set on the pod level and is applied to all the containers in the pod. So we cannot set --restart for containers in a pod.

Reply:
Yes, that's still a breaking change.  Existing deployments using `--pod new:foo --restart=...` will start to break.

We can follow K8s _by default_ but that doesn't mean we need to restrict users from changing individual policies, especially if it worked before.  Don't we already support individual restart policies when configured in K8s YAML?

I am sure, we can find a solution where containers will _inherit_ the restart policy from the pod unless the user explicitly overwrites it.

Reply:
So just thinking a step further for `podman kube generate`, if the pod has a restart policy set to `on failure` and there are two containers in the pod where one has restart policy set to `never` and the other has no restart policy set, do we inherit the restart policy of the pod, which would be `on failure`.

During the kube generate command, do we just ignore the individual container's restart policy and set the policy for the whole pod to `on failure`? 

and what about when the pod has no restart policy set and only one container in it has it set to `never`, do we apply a restart policy of `never` to the whole pod then when doing kube generate?

I understand that this is a breaking change, but being able to have different restart policies for a pod and for a container within the pod seems confusing, and may be confusing for users who are probably coming from the k8s world or for users who use pods in podman and when they transition to k8s, they notice the behavior is quite different there.

We can ensure that deployments that use `--pod=new:foo --restart` don't break and only restrict the use of `--restart` with `podman create/run` when the `--pod` flag doesn't have a `new:` prefix.

Reply:
> During the kube generate command, do we just ignore the individual container's restart policy and set the policy for the whole pod to on failure?

`generate kube` should reflect the restart policy of each container. If the pod has a restart policy set, this policy should bubble down to all containers by default unless explicitly overridden by the user.  So all containers will have policy set that `generate kube` can use.  If a container has no policy set, then its pod has none either.

> and what about when the pod has no restart policy set and only one container in it has it set to never, do we apply a restart policy of never to the whole pod then when doing kube generate?

An individual policy should only apply to the individual container.

> I understand that this is a breaking change, but being able to have different restart policies for a pod and for a container within the pod seems confusing, and may be confusing for users who are probably coming from the k8s world or for users who use pods in podman and when they transition to k8s, they notice the behavior is quite different there.

If users transition from K8s to Podman they will be using K8s YAML where we already allow to set individual restart policies for containers.

Reply:
In the k8s yaml, you can only set the restart policy at the pod level, which is applied to all the containers in the pod. There is no way to specify a restart policy for each container in a pod in a k8s yaml. When we do a `kube generate`, there will only be one restart policy value set in the yaml at the pod level. 

That is why I was trying to understand/figure out that if we allow setting the restart policy for a pod and allow setting separate restart policies for containers within a pod, how do we determine what the final restart policy would be set at the pod level when doing `kube generate`.

I do think having restart policy behave differently for a pod in k8s and in podman will be confusing to the user and this is more of a missing feature or bug instead of breaking change if we decide to only allow setting the restart policy at the pod level when creating containers in a pod.

If we continue to allow setting separate restart policy for containers in a pod, I think the yaml generated via `kube yaml` should always just pick the pod restart policy - that would be `Always` by default or what the user set it to during pod creation.

cc @mheon @baude @rhatdan

Reply:
> There is no way to specify a restart policy for each container in a pod in a k8s yaml. When we do a kube generate, there will only be one restart policy value set in the yaml at the pod level.

You are right, thanks for clarifying.

> I do think having restart policy behave differently for a pod in k8s and in podman will be confusing to the user and this is more of a missing feature or bug instead of breaking change if we decide to only allow setting the restart policy at the pod level when creating containers in a pod.

We must take backwards compatibility seriously.  It has been working so far and an unknown amount of existing workloads may break. K8s compat is important for `podman kube *` but, for me, it's not a strong argument to break `podman pod` behavior.  If it would be my workload, I'd be upset.

`podman pod` existed before `podman kube` and it's commonly used to bundle workloads together (e.g., easy communication via `localhost`).  IMO it is fundamentally different to K8s.

> If we continue to allow setting separate restart policy for containers in a pod, I think the yaml generated via kube yaml should always just pick the pod restart policy - that would be Always by default or what the user set it to during pod creation.

It's definitely something to be improved.  Looking at the current behavior, it seems like the last container's restart policy is set for the entire pod which seems extremely random.  Using the one of the pod seems like a good idea.

Reply:
I agree, we don't want to break anyone. I have removed the restriction and added a note that in podman v5.0, we should reconsider this.

This PR focuses on adding restart policy to podman pod only, so I will open a follow up PR once this is in to make the changes for kube gen and play.

Thank you for the reviews!